src/app/instance.cc

```diff
@@ -212,17 +186,67 @@ void Instance::SetupTunnels() {
       return;
     }
   }  // end of iteration block
-  if (m_IsReloading) {
-    LOG(info) << "Instance: " << client_count << " client tunnels updated";
-    LOG(info) << "Instance: " << server_count << " server tunnels updated";
-    // TODO(unassigned): this was never fully implemented
-    RemoveOldTunnels(updated_tunnels);
-    return;
-  }
   LOG(info) << "Instance: " << client_count << " client tunnels created";
   LOG(info) << "Instance: " << server_count << " server tunnels created";
 }
 
+void Instance::ReloadTunnels()
+{
+  // List of tunnels that exist after update
+  // TODO(unassigned): ensure that default IRC and eepsite tunnels aren't removed?
+  std::vector<std::string>
+      updated_tunnels;  // TODO(unassigned): this was never fully implemented
+  // Count number of tunnels
+  std::size_t client_count = 0, server_count = 0;
+  // Iterate through each section in tunnels config
+  for (auto const& tunnel : m_Config.GetParsedTunnelsConfig())
+    {
+      try
+        {
+          // Test which type of tunnel (client or server)
+          if (tunnel.type == GetConfig().GetAttribute(Key::Client)
+              || tunnel.type == GetConfig().GetAttribute(Key::IRC))
+            {  // TODO(unassigned): see #9
+              auto client_tunnel =
+                  kovri::client::context.GetClientTunnel(tunnel.port);
+              if (client_tunnel && client_tunnel->GetName() != tunnel.name)
+                {
+                  // Conflicting port
+                  // TODO(unassigned): what if we interchange two client tunnels' ports?
+                  // TODO(unassigned): the addresses could differ
+                  LOG(error) << "Instance: " << tunnel.name
+                             << " will not be updated, conflicting port";
+                  continue;
+                }
+              kovri::client::context.UpdateClientTunnel(tunnel);
+              ++client_count;
+            }
+          else
+            {
+              bool is_http =
+                  (tunnel.type == GetConfig().GetAttribute(Key::HTTP));
+              kovri::client::context.UpdateServerTunnel(tunnel, is_http);
+              ++server_count;
+            }
+        }
+      catch (const std::exception& ex)
+        {
+          LOG(error) << "Instance: exception during tunnel setup: "
+                     << ex.what();
+          return;
+        }
```

Comment:
You can get rid of this block and implement the exception dispatcher in the next block.